Win conditions

lpenzey opened the pull request.

lpenzey committed:
Define conditions for end of game

Status module generates every possible combination of moves that could
produce a win and uses those combinations to check if the status of the
game is over, tied, or won. The winning combinations are all built upon
first chunking the rows of the board. The rows are created by chunking
the board every n elements, where n is the square root of the size of
the board.

lpenzey committed:
Connect modules into playable game loop

The game is now playable between a human and a (dim) computer opponent.
Changes in this commit include moving board related functions from
Validity to Board and adding additional tests to increase test coverage.

rmulhol commented on .vscode/launch.json:
```diff
@@ -0,0 +1,29 @@
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
```
looks like you're already intending to get this out of here based on the changes to the `.gitignore` - 👍 to that. You probably want to do some sort of rebase and perhaps check out [git reset](https://git-scm.com/book/en/v2/Git-Tools-Reset-Demystified).

Probably also best to remove the binary while you're at it, since that'll be generated based on the OS.

rmulhol commented on lib/TTT/computer_player.ex:
```diff
@@ -15,7 +15,13 @@ defmodule ComputerPlayer do
+        select_move(state)
+        |> State.set_move(state)
+
+      select_move(state) == {:error, "empty error"} ->
```
why does this case need to be handled?

lpenzey replied:
This came about so that the computer_player wouldn't initiate setting a move if the board had already been filled, as it resulted in {:error, "empty error"} being passed to State.set_move.

rmulhol commented on lib/TTT/game.ex:
```diff
@@ -3,12 +3,20 @@ defmodule Game do
+    Output.end_of_game(state)
+  end
+
+  def play(false, state) do
```
Usually it's best to avoid passing booleans to functions, since it can indicate a violation of the single responsibility principle. Have you come across that tidbit in _Clean Code_? Think it applies here?

lpenzey replied:
Edit: passing a boolean here was a mistake, but I'm still uncertain about how to adhere to SRP with a recursive function that uses pattern matching to determine when the base case has been reached. Was the issue here strictly that I was passing true/false or is it that plus the fact that I have two definitions of play?

rmulhol commented on test/computer_player_test.exs:
```diff
@@ -40,6 +40,13 @@ defmodule ComputerPlayerTest do
+      computer_move_state = ComputerPlayer.make_move(context[:initial_state])
+      board_values = Map.values(computer_move_state.board)
+
+      assert Enum.find(board_values, fn x -> x == "X" end) == "X"
```
There's a lot going on in this test to assert that one space on the board is "X". I wonder if there might be a different way to structure the code to make it easier to test.

rmulhol reviewed (commented):
Looking good!

lpenzey committed:
Remove binary and update gitignore

lpenzey committed:
Rename Board to ConsoleBoardPresenter

Board was formerly concerned with both the storage mechanism of moves
and how the board is rendered and displayed in the console. Board has
been renamed to ConsoleBoardPresenter and is strictly focused on
rendering and displaying the board in the console.

The Functions that relate to the storage mechanism for moves have been
relocated to State module.

lpenzey committed:
Remove flag arguments from play function

Play now accepts a single argument (state) and the end of the game is
managed by a conditional that checks if the game is over or if it is
still in progress which helps Game.play adhere to the single
responsibility principle.

lpenzey committed:
Inject IO module as dependency to Input

The Input and Output modules now receive the built-in IO module as an injected
dependency that's retrieved from the config file.

lpenzey committed:
Add mock to test input analysis
